File: 3.无重复字符的最长子串.java
import java.util.HashMap;

/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> window = new HashMap<>();
        int left = 0, right = 0;
        int res = 0;
        while (right < s.length()) {
            char r = s.charAt(right);
            right++;
            window.put(r, window.getOrDefault(r, 0) + 1);
            while (window.getOrDefault(r, 0) > 1) {
                char l = s.charAt(left);
                left++;
                window.put(l, window.get(l) - 1);
            }
            if (res < right - left) {
                res = right - left;
            }
        }
        return res;
    }
}
// @lc code=end

File: python/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            return 0
            
        # 使用字典存储字符最后一次出现的索引
        char_map = {}
        left = 0
        max_len = 0
        
        for right in range(len(s)):
            # 如果当前字符已经在窗口内出现过
            if s[right] in char_map and char_map[s[right]] >= left:
                # 将左边界直接跳到重复字符的下一位
                left = char_map[s[right]] + 1
            
            # 更新当前字符的位置
            char_map[s[right]] = right
            # 计算当前窗口长度并更新最大值
            max_len = max(max_len, right - left + 1)
            
        return max_len
        
        
        
# @lc code=end

s = "pwwkew"
print(Solution().lengthOfLongestSubstring(s))